tools/SourceKit/tools/sourcekitd/bin/XPC/Client/sourcekitd.cpp

```diff
@@ -285,9 +288,27 @@ static void handleInternalUIDRequest(xpc_object_t XVal,
 
 static void handleInterruptedConnection(xpc_object_t event, xpc_connection_t conn);
 
+extern "C" const char __dso_handle[];
+
 static void initializeXPCClient() {
   assert(!GlobalConn);
-  GlobalConn = xpc_connection_create(SOURCEKIT_XPCSERVICE_IDENTIFIER, nullptr);
+
+  Dl_info dlinfo;
+  dladdr(__dso_handle, &dlinfo);
+
+  // '.../usr/lib/sourcekitd.framework/sourcekitd'
+  llvm::SmallString<128> serviceNamePath(dlinfo.dli_fname);
```

Comment:
Not sure if it's possible for `dladdr` to fail here, but could check for an empty path here to make sure we don't try to read `<cwd>/Resources/xpc_service_name.txt`. Not sure if it matter though

Reply:
Not sure either, but better to be safe than sorry! Updated, thanks.